File: python/benchmarks/bench_utils.py
#!/usr/bin/env python3
# Copyright 2022 Lance Developers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import os
import pathlib
from abc import ABC, abstractmethod
from functools import wraps
import multiprocessing as mp
from typing import Iterable

import click
import pandas as pd
import time

import pyarrow as pa
import pyarrow.fs
import pyarrow.dataset as ds
import pyarrow.parquet as pq

import lance

__all__ = ["download_uris", "timeit", "get_dataset", "get_uri", "BenchmarkSuite"]

KNOWN_FORMATS = ["lance", "parquet", "raw"]


def read_file(uri) -> bytes:
    fs, key = pyarrow.fs.FileSystem.from_uri(uri)
    return fs.open_input_file(key).read()


def download_uris(uris: Iterable[str], func=read_file) -> Iterable[bytes]:
    if isinstance(uris, pd.Series):
        uris = uris.values
    pool = mp.Pool(mp.cpu_count() - 1)
    data = pool.map(func, uris)
    return data


def timeit(func):
    @wraps(func)
    def timeit_wrapper(*args, **kwargs):
        start_time = time.perf_counter()
        result = func(*args, **kwargs)
        end_time = time.perf_counter()
        total_time = end_time - start_time
        # first item in the args, ie `args[0]` is `self`
        print(f"Function {func.__name__}{args} {kwargs} Took {total_time:.4f} seconds")
        return result

    return timeit_wrapper


def get_dataset(uri: str) -> ds.Dataset:
    """
    Return a pyarrow Dataset stored at the given uri
    """
    if uri.endswith(".lance"):
        return lance.dataset(uri)
    return ds.dataset(uri)


def get_uri(base_uri: str, dataset_name: str, fmt: str, flavor: str = None) -> str:
    """
    Return the uri to the dataset with the given specifications

    Parameters
    ----------
    base_uri: str
        Base uri to the root of the benchmark dataset catalog
    dataset_name: str
        Catalog name of the dataset (e.g., coco, oxford_pet)
    fmt: str
        'lance', 'parquet', or 'raw'
    flavor: str, optional
        We may store different flavors for parquet and lance,
        e.g., with image links but not bytes
    """
    return f"{base_uri}/{dataset_name}{('_' + flavor) if flavor else ''}.{fmt}"


class BenchmarkSuite:
    def __init__(self, name: str):
        self.name = name
        self._benchmarks = {}
        self._results = {}

    def benchmark(self, name, key=None):
        def decorator(func):
            b = Benchmark(name, func, key=key)
            self._benchmarks[name] = b
            return func

        return decorator

    def get_benchmark(self, name):
        return self._benchmarks[name]

    def list_benchmarks(self):
        return self._benchmarks.values()

    def create_main(self):
        @click.command
        @click.option(
            "-u",
            "--base-uri",
            required=True,
            type=str,
            help="Base uri to the benchmark dataset catalog",
        )
        @click.option(
            "-f", "--format", "fmt", help="'lance', 'parquet', or 'raw'. Omit for all"
        )
        @click.option(
            "--flavor",
            type=str,
            help="external if parquet/lance had external images version",
        )
        @click.option(
            "-b", "--benchmark", type=str, help="which benchmark to run. Omit for all"
        )
        @click.option(
            "-r", "--repeats", type=int, help="number of times to run each benchmark"
        )
        @click.option(
            "-o", "--output", type=str, help="save timing results to directory"
        )
        def main(base_uri, fmt, flavor, benchmark, repeats, output):
            if fmt:
                fmt = fmt.strip().lower()
                assert fmt in KNOWN_FORMATS
                fmt = [fmt]
            else:
                fmt = KNOWN_FORMATS
            base_uri = f"{base_uri}/datasets/{self.name}"

            def run_benchmark(bmark):
                b = bmark.repeat(repeats or 1)
                for f in fmt:
                    b.run(base_uri=base_uri, fmt=f, flavor=flavor)
                if output:
                    path = pathlib.Path(output) / f"{bmark.name}.csv"
                    b.to_df().to_csv(path, index=False)

            if benchmark is not None:
                b = self.get_benchmark(benchmark)
                run_benchmark(b)
            else:
                [run_benchmark(b) for b in self.list_benchmarks()]

        return main


class Benchmark:
    def __init__(self, name, func, key=None, num_runs=1):
        self.name = name
        self.func = func
        self.key = key
        self.num_runs = num_runs
        self._timings = {}

    def repeat(self, num_runs: int):
        return Benchmark(self.name, self.func, key=self.key, num_runs=num_runs)

    def run(self, *args, **kwargs):
        output = None
        func = self.timeit("total")(self.func)
        for i in range(self.num_runs):
            output = func(*args, **kwargs)
        return output

    def to_df(self):
        return pd.DataFrame(self._timings)

    def timeit(self, name):
        def benchmark_decorator(func):
            @wraps(func)
            def timeit_wrapper(*args, **kwargs):
                start_time = time.perf_counter()
                result = func(*args, **kwargs)
                end_time = time.perf_counter()
                total_time = end_time - start_time
                # first item in the args, ie `args[0]` is `self`
                print(
                    f"Function {func.__name__}{args} {kwargs} Took {total_time:.4f} seconds"
                )
                key = tuple([name] + [kwargs.get(k) for k in self.key])
                self._timings.setdefault(key, []).append(total_time)
                return result

            return timeit_wrapper

        return benchmark_decorator


class DatasetConverter(ABC):
    def __init__(self, name, uri_root):
        self.name = name
        self.uri_root = uri_root

    @abstractmethod
    def read_metadata(self) -> pd.DataFrame:
        pass

    def default_dataset_path(self, fmt, flavor=None):
        suffix = f"_{flavor}" if flavor else ""
        return os.path.join(self.uri_root, f"{self.name}{suffix}.{fmt}")

    def save_df(self, df, fmt="lance", output_path=None):
        output_path = output_path or self.default_dataset_path(fmt, "links")
        table = pa.Table.from_pandas(df, self.get_schema())
        if fmt == "parquet":
            pq.write_table(table, output_path)
        elif fmt == "lance":
            lance.write_table(table, output_path)
        return table

    @abstractmethod
    def image_uris(self, table):
        pass

    def make_embedded_dataset(
        self, table: pa.Table, fmt="lance", output_path=None, **kwargs
    ):
        output_path = output_path or self.default_dataset_path(fmt)
        uris = self.image_uris(table)
        images = download_uris(pd.Series(uris))
        arr = pa.BinaryArray.from_pandas(images)
        embedded = table.append_column(pa.field("image", pa.binary()), arr)
        if fmt == "parquet":
            pq.write_table(embedded, output_path, **kwargs)
        elif fmt == "lance":
            lance.write_table(embedded, output_path, **kwargs)
        return embedded

    @abstractmethod
    def get_schema(self):
        pass

File: cpp/benchmarks/bench_utils.cc
//  Copyright 2022 Lance Authors
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#include "bench_utils.h"

#include <arrow/dataset/discovery.h>
#include <arrow/dataset/file_parquet.h>
#include <arrow/io/api.h>
#include <arrow/util/string.h>
#include <fmt/format.h>
#include <lance/arrow/file_lance.h>

#include <filesystem>
#include <future>

std::shared_ptr<::arrow::dataset::Scanner> OpenScanner(
    const std::string& uri,
    const std::vector<std::string>& columns,
    std::optional<arrow::compute::Expression> filter,
    std::optional<int> batch_size) {
  auto format = std::shared_ptr<arrow::dataset::FileFormat>();
  if (uri.ends_with(".lance")) {
    format = lance::arrow::LanceFileFormat::Make();
  } else if (uri.ends_with(".parquet") || uri.ends_with(".parq")) {
    format.reset(new arrow::dataset::ParquetFileFormat());
  }
  auto factory = ::arrow::dataset::FileSystemDatasetFactory::Make(
                     uri, format, arrow::dataset::FileSystemFactoryOptions())
                     .ValueOrDie();
  auto dataset = factory->Finish().ValueOrDie();
  // fmt::print("{} dataset: {} groups={}\n", format->type_name(), uri,
  // dataset->schema()->ToString());
  auto scan_builder = dataset->NewScan().ValueOrDie();
  if (batch_size.has_value()) {
    fmt::print("Setting batch size: {}\n", batch_size.value());
    assert(scan_builder->BatchSize(batch_size.value()).ok());
  }
  if (filter.has_value()) {
    assert(scan_builder->Filter(filter.value()).ok());
  }
  assert(scan_builder->Project(columns).ok());
  assert(scan_builder->UseThreads().ok());
  auto scanner = scan_builder->Finish().ValueOrDie();
  /// TODO: somehow scanner builder does not honor batch size??
  scanner->options()->batch_size = batch_size.value();
  if (format->type_name() == "lance") {
    /// set to 16 will crash EC2 instance.
    scanner->options()->batch_readahead = 8;
  }
  return scan_builder->Finish().ValueOrDie();
}

std::shared_ptr<::arrow::io::RandomAccessFile> OpenUri(const std::string& uri, bool ignore_error) {
  if (uri.starts_with("s3")) {
    auto fs = arrow::fs::FileSystemFromUriOrPath(uri).ValueOrDie();
    auto file = fs->OpenInputFile(uri.substr(std::string("s3://").size()));
    if (!file.ok() && ignore_error) {
      return nullptr;
    }
    return file.ValueOrDie();
  } else {
    auto path = std::filesystem::absolute(uri);
    auto fs = arrow::fs::FileSystemFromUriOrPath(path).ValueOrDie();
    auto file = fs->OpenInputFile(path);
    if (!file.ok() && ignore_error) {
      return nullptr;
    }
    return file.ValueOrDie();
  }
}

/// Open Dataset from the URI.
std::shared_ptr<::arrow::dataset::FileSystemDataset> OpenDataset(const std::string& uri,
                                                                 const std::string& format) {
  std::string path;
  auto fs = ::arrow::fs::FileSystemFromUriOrPath(uri, &path).ValueOrDie();
  std::shared_ptr<arrow::dataset::FileFormat> file_format;
  if (format == "lance") {
    file_format.reset(new lance::arrow::LanceFileFormat());
  } else if (format == "parquet") {
    file_format.reset(new arrow::dataset::ParquetFileFormat());
  } else {
    fmt::print(stderr, "Unsupported file format: {}\n", format);
    assert(false);
  }
  auto selector = ::arrow::fs::FileSelector();
  selector.base_dir = path;
  selector.recursive = true;
  auto factory = arrow::dataset::FileSystemDatasetFactory::Make(
                     fs, selector, file_format, arrow::dataset::FileSystemFactoryOptions())
                     .ValueOrDie();
  auto dataset = factory->Finish().ValueOrDie();
  return std::dynamic_pointer_cast<::arrow::dataset::FileSystemDataset>(dataset);
}

void ReadAll(const std::string& uri, bool ignore_error) {
  auto infile = OpenUri(uri, ignore_error);
  if (infile) {
    auto buf = infile->Read(infile->GetSize().ValueOrDie()).ValueOrDie();
  }
}

void ReadAll(const std::vector<std::string>& uris, size_t num_workers, bool ignore_error) {
  auto total = uris.size();
  for (decltype(total) start = 0; start < total; start += num_workers) {
    std::vector<std::future<void>> futures;
    auto len = std::min(num_workers, total - start);
    for (size_t i = 0; i < len; i++) {
      auto& uri = uris[start + i];
      futures.emplace_back(std::async(
          std::launch::async,
          [](const std::string& uri, bool e) {
            auto infile = OpenUri(uri, e);
            if (infile) {
              auto buf = infile->Read(infile->GetSize().ValueOrDie()).ValueOrDie();
            }
            return;
          },
          uri,
          ignore_error));
    }
    for (auto& f : futures) {
      f.wait();
    }
  }
}